feat: make it possible to deactivate CR validation

metacosm opened the pull request:
Fixes #354

adam-sandor commented:
I know it's a bit late for this... but do we ever need to validate the CRD? Users would find out it's not there when they try to create CRs, with a standard K8s error message. The operator can totally ignore if the CRD exists it only cares about getting events about CRs.

metacosm commented:
The error message you get is quite cryptic, which is why I added the functionality in the first place. @daniellavoie also added additional validation to pick up on missing `implements Namespaced` as well since these two things are common "beginners" mistakes.

adam-sandor commented on operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/config/ConfigurationService.java:
```diff
@@ -42,4 +42,17 @@ default Config getClientConfiguration() {
+   *
+   * @return {@code true} if CRDs should be checked (default), {@code false} otherwise
+   */
+  default boolean validateCustomResources() {
```
It should be validateCustomResourceDefinitions

metacosm replied:
Not really, we're validating the Custom Resources from the CRD: if the CRD has been accepted by the cluster, it's presumed to be valid. However, the `CustomResource` implementation you want to use in your controller might not match what the CRD defines so it's really validating that these classes match what the cluster knows and wants.

adam-sandor replied:
ah ok, got it! then ignore my other comments too and go ahead to merge the PR :)

adam-sandor commented on operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/config/Utils.java:
```diff
@@ -12,6 +12,7 @@
 public class Utils {
 
   private static final Logger log = LoggerFactory.getLogger(Utils.class);
+  public static final String VALIDATE_CR_ENV_KEY = "JAVA_OPERATOR_SDK_VALIDATE_CR";
```
should be ...VALIDATE_CRD...

adam-sandor commented on operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/config/Utils.java:
```diff
@@ -48,4 +49,13 @@ public static Version loadFromProperties() {
         builtTime);
   }
+
+  public static boolean isValidateCustomResourcesEnvVarSet() {
```
again.. customResourceDefinition instead of customResource. I know it's long but it's confusing otherwise

adam-sandor commented on operator-framework-quarkus-extension/deployment/src/main/java/io/javaoperatorsdk/quarkus/extension/deployment/QuarkusExtensionProcessor.java:
```diff
@@ -81,11 +81,16 @@ void createConfigurationServiceAndOperator(
             .collect(Collectors.toList());
 
     final var version = Utils.loadFromProperties();
+    final var validateCustomResources =
```
cr -> crd

adam-sandor commented on operator-framework-quarkus-extension/runtime/src/main/java/io/javaoperatorsdk/quarkus/extension/ConfigurationServiceRecorder.java:
```diff
@@ -13,9 +13,14 @@
-      Version version, List<ControllerConfiguration> controllerConfigs) {
+      Version version,
+      List<ControllerConfiguration> controllerConfigs,
+      boolean validateCustomResources) {
```
cr -> crd

adam-sandor commented on operator-framework-quarkus-extension/runtime/src/main/java/io/javaoperatorsdk/quarkus/extension/ExternalConfiguration.java:
```diff
@@ -1,13 +1,22 @@
+   * registering the associated controller.
+   */
+  @ConfigItem(defaultValue = "true")
+  public Optional<Boolean> validateCustomResources;
```
cr -> crd

adam-sandor commented on operator-framework-quarkus-extension/runtime/src/main/java/io/javaoperatorsdk/quarkus/extension/QuarkusConfigurationService.java:
```diff
@@ -13,14 +13,19 @@
+      Version version,
+      List<ControllerConfiguration> configurations,
+      KubernetesClient client,
+      boolean validateCustomResources) {
```
cr -> crd

adam-sandor commented on operator-framework-quarkus-extension/runtime/src/main/java/io/javaoperatorsdk/quarkus/extension/QuarkusConfigurationService.java:
```diff
@@ -35,6 +40,11 @@ public Config getClientConfiguration() {
   }
 
+  @Override
+  public boolean validateCustomResources() {
```
cr -> crd

adam-sandor commented on operator-framework-quarkus-extension/tests/src/main/java/io/javaoperatorsdk/quarkus/it/TestOperatorApp.java:
```diff
@@ -18,6 +18,12 @@
 
+  @GET
+  @Path("validateCR")
+  public boolean validateCR() {
```
validateCRD

adam-sandor commented on operator-framework-quarkus-extension/tests/src/main/resources/application.properties:
```diff
@@ -1,2 +1,3 @@
-quarkus.operator-sdk.controllers.annotation.namespaces=bar
\ No newline at end of file
+quarkus.operator-sdk.controllers.annotation.namespaces=bar
+quarkus.operator-sdk.validate-custom-resources=false
```
validate-custom-resource-definitions

adam-sandor commented on operator-framework-quarkus-extension/tests/src/test/java/io/javaoperatorsdk/quarkus/it/QuarkusExtensionProcessorTest.java:
```diff
@@ -18,6 +18,11 @@
 
+  @Test
+  void shouldNotValidateCRs() {
+    given().when().get("/operator/validateCR").then().statusCode(200).body(is("false"));
```
validateCRD

adam-sandor commented on operator-framework/src/main/java/io/javaoperatorsdk/operator/config/runtime/DefaultConfigurationService.java:
```diff
@@ -36,4 +36,9 @@ public static ConfigurationService instance() {
   }
+
+  @Override
+  public boolean validateCustomResources() {
```
cr -> crd

adam-sandor reviewed (changes requested):
I think there are a lot of places where CRD should replace CR. If there was a good reason to use CR instead then I missed it...

metacosm commented:
The methods/fields are renamed in #361 to address potential confusion on validation scope.
